crypto/bn/bn_prime.c

```diff
@@ -49,15 +49,91 @@ int BN_GENCB_call(BN_GENCB *cb, int a, int b)
     return 0;
 }
 
-int BN_generate_prime_ex(BIGNUM *ret, int bits, int safe,
-                         const BIGNUM *add, const BIGNUM *rem, BN_GENCB *cb)
+/*
+ * This defines the security level |sec_level| in bits of doing |mr_checks|
+ * Miller-Rabin on a random prime of |prime_size|.
+ *
+ * The table is sorted by |mr_checks|, then |prime_size|. This is so that we
+ * find the minimum number of Miller-Rabin for a given |prime_size| and
+ * |sec_level|.
+ *
+ * This table is generated using the algorithm of FIPS 186-4 section F.1
+ */
+static const struct {
+    int prime_size;
+    int mr_checks;
+    int sec_level; } prime_mr_sec_level[] = {
+    { 512, 4, 73 },
+    { 768, 4, 93 },
+    { 1024, 4, 109 },
+    { 2048, 4, 161 },
+    { 1536, 4, 137 },
+    { 3072, 4, 201 },
+    { 4096, 4, 235 },
+    { 6144, 4, 292 },
+    { 8192, 4, 340 },
+    { 11264, 4, 402 },
+    { 512, 5, 83 },
+    { 768, 5, 105 },
+    { 1024, 5, 124 },
+    { 1536, 5, 155 },
+    { 2048, 5, 182 },
+    { 3072, 5, 227 },
+    { 4096, 5, 264 },
+    { 6144, 5, 328 },
+    { 8192, 5, 382 },
+    { 11264, 5, 451 },
+    { 512, 6, 93 },
+    { 768, 6, 116 },
+    { 1024, 6, 137 },
+    { 1536, 6, 171 },
+    { 2048, 6, 200 },
+    { 3072, 6, 249 },
+    { 4096, 6, 291 },
+    { 6144, 6, 361 },
+    { 8192, 6, 420 },
+    { 11264, 6, 496 }};
+
+/*
+ * Returns the number of Miller-Rabin tests that needs to be performed on
+ * the prime of size |prime_size| bits to reach the security level of
+ * |security_bits|.
+ */
+int bn_prime_checks_for_size_ex(int prime_size, int security_bits)
+{
+    size_t i;
+
+    for (i = 0; i < OSSL_NELEM(prime_mr_sec_level); i++)
+        if (prime_mr_sec_level[i].prime_size <= prime_size &&
+            prime_mr_sec_level[i].sec_level >= security_bits)
+            return prime_mr_sec_level[i].mr_checks;
+
+    /*
+     * No pre-generated results. The Handbook of Applied Cryptography, in 4.47
+     * gives this as upper bound.
+     */
+    if (prime_size > 60)
+        return (security_bits + 1) / 2;
+
+    /* Can't find any good numbers for this. */
+    return security_bits * 2; 
+}
+
+/*
+ * BN_generate_prime_ex() with an extra security_bits parameter
```

Comment:
The extra parameter seems to be the number of checks, not the number of security bits.